src/plugins/expressions/common/expression_functions/specs/moving_average.ts

```diff
@@ -0,0 +1,156 @@
+  type: 'datatable',
+
+  inputTypes: ['datatable'],
+
+  help: i18n.translate('expressions.functions.movingAverage.help', {
+    defaultMessage: 'Calculates the moving average of a column in a data table',
+  }),
+
+  args: {
+    by: {
+      help: i18n.translate('expressions.functions.movingAverage.args.byHelpText', {
+        defaultMessage: 'Column to split the moving average calculation by',
+      }),
+      multi: true,
+      types: ['string'],
+      required: false,
+    },
+    inputColumnId: {
+      help: i18n.translate('expressions.functions.movingAverage.args.inputColumnIdHelpText', {
+        defaultMessage: 'Column to calculate the moving average of',
+      }),
+      types: ['string'],
+      required: true,
+    },
+    outputColumnId: {
+      help: i18n.translate('expressions.functions.movingAverage.args.outputColumnIdHelpText', {
+        defaultMessage: 'Column to store the resulting moving average in',
+      }),
+      types: ['string'],
+      required: true,
+    },
+    outputColumnName: {
+      help: i18n.translate('expressions.functions.movingAverage.args.outputColumnNameHelpText', {
+        defaultMessage: 'Name of the column to store the resulting moving average in',
+      }),
+      types: ['string'],
+      required: false,
+    },
+    window: {
+      help: i18n.translate('expressions.functions.movingAverage.args.windowHelpText', {
+        defaultMessage: 'The size of window to "slide" across the histogram.',
+      }),
+      types: ['number'],
+      required: false,
+      default: 5,
+    },
+  },
+
+  fn(input, { by, inputColumnId, outputColumnId, outputColumnName, window = 5 }) {
+    const resultColumns = buildResultColumns(
+      input,
+      outputColumnId,
+      inputColumnId,
+      outputColumnName
+    );
+
+    if (!resultColumns) {
+      return input;
+    }
+
+    const lastNValuesByBucket: Partial<Record<string, number[]>> = {};
+    return {
+      ...input,
+      columns: resultColumns,
+      rows: input.rows.map((row) => {
+        const newRow = { ...row };
+        const bucketIdentifier = getBucketIdentifier(row, by);
+        const lastNValues = lastNValuesByBucket[bucketIdentifier];
+        const currentValue = newRow[inputColumnId];
+        if (lastNValues != null && lastNValues.length && currentValue != null) {
+          const sum = lastNValues.reduce((acc, current) => acc + current, 0);
+          newRow[outputColumnId] = sum / lastNValues.length;
+        } else {
+          newRow[outputColumnId] = undefined;
+        }
+
+        if (currentValue != null && lastNValues != null) {
+          lastNValuesByBucket[bucketIdentifier] = [...lastNValues, Number(currentValue)].slice(
+            -window
+          );
+        } else if (currentValue != null) {
+          lastNValuesByBucket[bucketIdentifier] = [Number(currentValue)];
+        } else if (currentValue == null && lastNValues == null) {
+          lastNValuesByBucket[bucketIdentifier] = undefined;
```

Comment:
nit: Seems like this branch can be removed without changing the behavior (it's setting `lastNValues` for the current bucket to undefined if it's already undefined)